bob/src/hw_metrics_collector.rs

```diff
@@ -179,53 +183,165 @@ enum CommandError {
-                    if let Some(iowait_str) = values.nth(3) {
-                        if let Ok(iowait) = Self::parse_iostat_result(iowait_str) {
-                            return Ok(iowait);
+        const CPU_IOWAIT_COLUMN: usize = 5;
+        let mut err = None;
+        match Self::stat_cpu_line() {
+            Ok(line) => {
+                let parts: Vec<&str> = line.split_whitespace().collect();
+                if parts.len() > CPU_IOWAIT_COLUMN {
+                    let mut sum = 0.;
+                    let mut f_iowait = 0.;
+                    for i in 1..parts.len() {
+                        match parts[i].parse::<f64>() {
+                            Ok(val) => {
+                                sum += val;
+                                if i == CPU_IOWAIT_COLUMN {
+                                    f_iowait = val;
+                                }
+                            },
+                            Err(_) => {
+                                let msg = format!("Can't parse {}", CPU_STAT_FILE);
+                                err = Some(msg);
+                                break;
+                            }
                         }
                     }
+                    if err.is_none() {
+                        return Ok(f_iowait * 100. / sum);
+                    }
+                } else {
+                    let msg = format!("CPU stat format in {} changed", CPU_STAT_FILE);
+                    err = Some(msg);
                 }
-            }
+            },
             Err(e) => {
-                self.iostat_avl = false;
-                return Err(CommandError::Primary(e));
+                err = Some(e);
+            }
+        }
+        self.procfs_avl = false;
+        Err(CommandError::Primary(err.unwrap()))
+    }
+}
+
+const DIFFCONTAINER_THRESHOLD: Duration = Duration::from_millis(500);
+
+struct TimeContainer {
+    last: Option<Instant>,
+    prelast: Option<Instant>,
+}
+
+enum TimeMoment {
+    Last(Duration),
+    Prelast(Duration),
+}
+
+impl TimeMoment {
+    fn elapsed_ms(&self) -> f64 {
+        match self {
+            TimeMoment::Last(dur) => dur.as_millis() as f64,
+            TimeMoment::Prelast(dur) => dur.as_millis() as f64,
+        }
+    }
+}
+
+impl TimeContainer {
+    fn new() -> Self {
+        Self {
+            last: None,
+            prelast: None,
+        }
+    }
+
+    fn update(&mut self, now: Instant) {
+        self.prelast = self.last;
+        self.last = Some(now);
+    }
+
+    fn tick(&mut self) -> Option<TimeMoment> {
+        let now = Instant::now();
+        if let Some(last) = self.last {
+            let elapsed = now.duration_since(last);
+            if elapsed > DIFFCONTAINER_THRESHOLD {
+                self.update(now);
+                return Some(TimeMoment::Last(elapsed));
+            } else if let Some(prelast) = self.prelast {
+                let elapsed = now.duration_since(prelast);
+                self.update(now);
```

Comment:
We don't need to update the time here because it could lead to a situation where the time elapsed since the prelast is less than `DIFFCONTAINER_THRESHOLD`.